modules/images/image-loading-optimization/storage/rest-api.php

```diff
@@ -0,0 +1,193 @@
+<?php
+/**
+ * REST API integration for the module.
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+const ILO_REST_API_NAMESPACE = 'image-loading-optimization/v1';
+
+const ILO_PAGE_METRICS_ROUTE = '/page-metrics';
```

Comment:
See above, if we changed the post type name, this name should be changed accordingly.

Additionally though, following up on https://github.com/WordPress/performance/pull/878#discussion_r1387484165, I don't think just this name is semantically correct. `POST /page-metrics` would be intended to create a new resource of this type, which isn't really the case as this endpoint doesn't create a new post of the metrics post type, but modifies an existing one with specific data for a request. That is effectively a custom method, so it shouldn't use just the resource based identifier.

Core's endpoints indeed do not include `:` characters, though it's probably not an issue introducing one. But if we want to tie more nicely into core's naming, we should simply use another forward slash to separate the custom method from the resource, i.e. `/page-metrics/store` (or `/url-metrics/store`). There are already a few endpoints in core that use a custom method in this way.

Maybe the `store` is also not super clear, how about `store-request`?

Reply:
> `POST /page-metrics` would be intended to create a new resource of this type, which isn't really the case as this endpoint doesn't create a new post of the metrics post type, but modifies an existing one with specific data for a request.

Actually, it can create a new resource of that type. If there were no page metrics gathered for the URL, then a new `ilo_page_metrics` post would be created. However, if the post does already exist, then essentially it creates a new _sub_-resource (by adding the page metric to the collection of page metrics for a given URL).

I don't feel strongly about the endpoint name, however. If you want to add a `:` then I'm fine with it. But IMO the `POST` method's broad semantics don't require it. In particular, the [HTTP spec](https://httpwg.org/specs/rfc7231.html#rfc.section.4.3.3) includes this use case for `POST`:

> Appending data to a resource's existing representation(s).

Reply:
Right, it _can_ create a new resource, but it doesn't allow, and that's not the purpose of the endpoint.

Regarding "Appending data to a resource's existing representation(s).", that applies to a specific resource. In other words, `POST /metrics/<id>` can append data to that resource. However `POST /metrics` does not touch any existing resource, it's typically intended to create a new resource. This is applied in most core endpoints in that way and also the best practice per https://google.aip.dev/133.

I don't feel strongly about the `:`, however I do think a custom method is the appropriate solution for what this endpoint does. It's not a distinct CRUD operation for a resource (as it can either create or update), and the parameters aren't fields on the resource.

If we consider the requests within a post of the post type a sub resource, we could go with that approach, but that would require knowing the post's identifier, which is probably too much of a hassle. So with the way things are now I think `POST /{page|url}-metrics{:|/}store-request` works best.

Reply:
I've implemented this in 698d89f. I think `store` is better than `store-request`, as it is immediately following `url-metrics`.

Reply:
I was thinking `store-request` because the data that is processed by that endpoint is from a single specific request.  And it's storing that within a URL metrics resource, rather than the URL metrics resource itself.

With only `store`, it sounds like you're storing a resource of the URL metrics post type, which if that was the case should not be a custom method. So I think `store-request` makes that more clear.

But not a blocker if you feel strongly about it.